src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.cs

```diff
@@ -384,5 +387,42 @@ public IEnumerable<string> GetParameterNamesForMethod(MethodDesc method)
                 yield return ecmaMethod.MetadataReader.GetString(p.Name);
             }
         }
+
+        public bool IsTypeDefinedExternally(TypeDesc type)
+        {
+            if (!_multifileCompilation)
+                return false;
+
+            if (type.ContainsGenericVariables)
+                return false;
+
+            if (!(type is MetadataType))
+                return false;
+
+            MetadataType metadataType = type as MetadataType;
```

Comment:
Nit: if you swap this line with the `if` above, you can make the `if` a simple null check which is easier to read than "not true that type is MetadataType".

Also, if you make it a cast to EcmaType, you can access the associated EcmaModule through the EcmaModule property, getting rid of the if and cast below.